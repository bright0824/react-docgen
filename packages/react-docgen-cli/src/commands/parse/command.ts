#!/usr/bin/env node
import glob from 'fast-glob';
import debugFactory from 'debug';
import { program } from 'commander';
import { builtinHandlers, parse } from 'react-docgen';
import { readFile } from 'fs/promises';
import outputResult from './output/outputResult.js';
import loadOptions from './options/loadOptions.js';
import outputError from './output/outputError.js';
import { resolve } from 'path';
import slash from 'slash';

const debug = debugFactory('react-docgen:cli');

const defaultIgnoreGlobs = [
  '**/node_modules/**',
  '**/__tests__/**',
  '**/__mocks__/**',
];

const defaultHandlers = Object.keys(builtinHandlers);

function collect(value: string, previous: string[]) {
  if (
    !previous ||
    previous === defaultIgnoreGlobs ||
    previous === defaultHandlers
  ) {
    previous = [];
  }

  const values = value.split(',');

  return previous.concat(values);
}

interface CLIOptions {
  defaultIgnores: boolean;
  failOnWarning: boolean;
  handlers?: string[];
  ignore: string[];
  importer?: string;
  out?: string;
  pretty: boolean;
  resolver?: string;
}

program
  .name('react-docgen-parse')
  .description(
    'Extract meta information from React components.\n' +
      'Either specify a paths to files or a glob pattern that matches multiple files.',
  )
  .option(
    '-o, --out <file>',
    'Store extracted information in the specified file instead of printing to stdout. If the file exists it will be overwritten.',
  )
  .option(
    '-i, --ignore <glob>',
    'Comma separated list of glob patterns which will ignore the paths that match. Can also be used multiple times.',
    collect,
    defaultIgnoreGlobs,
  )
  .option(
    '--no-default-ignores',
    'Do not ignore the node_modules, __tests__, and __mocks__ directories.',
  )
  .option('--pretty', 'Print the output JSON pretty', false)
  .option(
    '--failOnWarning',
    'Fail with exit code 2 on react-docgen component warnings. This includes "no component found" and "multiple components found" warnings.',
    false,
  )
  .option(
    '--resolver <resolver>',
    'Built-in resolver name (findAllComponentDefinitions, findAllExportedComponentDefinitions, findExportedComponentDefinition) or path to a module that exports a resolver.',
    'findExportedComponentDefinition',
  )
  .option(
    '--importer <importer>',
    'Built-in importer name (fsImport, ignoreImporter) or path to a module that exports an importer.',
    'fsImporter',
  )
  .option(
    '--handlers <handlers>',
    'Comma separated list of handlers to use. Can also be used multiple times. When used no default handlers will be added.',
    collect,
    defaultHandlers,
  )
  .argument('<globs...>', 'Can be globs or paths to files')
  .action(async (globs: string[], input: CLIOptions) => {
    const {
      defaultIgnores,
      failOnWarning,
      handlers,
      ignore,
      importer,
      out: output,
      pretty,
      resolver,
    } = input;

    let finalIgnores = ignore;

    // Push the default ignores unless the --no-default-ignore is set
    if (defaultIgnores === true && ignore !== defaultIgnoreGlobs) {
      finalIgnores.push(...defaultIgnoreGlobs);
    } else if (defaultIgnores === false && ignore === defaultIgnoreGlobs) {
      finalIgnores = [];
    }

    const options = await loadOptions({
      handlers,
      importer,
      resolver,
    });
    // we use slash to convert windows backslashes to unix format so fast-glob works
    const files = await glob(globs.map(slash), {
      ignore: finalIgnores?.map(ignorePath => {
        ignorePath = ignorePath.trim();
        // If the ignore glob starts with a dot we need to resolve the path to an
        // absolute path in order for it to work
        if (ignorePath.startsWith('.')) {
          ignorePath = resolve(process.cwd(), ignorePath);
        }

        // we use slash to convert windows backslashes to unix format so fast-glob works
        return slash(ignorePath);
      }),
    });
    const result = {};
    let errorEncountered = false;

    await Promise.all(
      files.map(async path => {
        debug(`Reading file ${path}`);
        const content = await readFile(path, 'utf-8');

        try {
          result[path] = parse(content, {
            filename: path,
            handlers: options.handlers,
            importer: options.importer,
            resolver: options.resolver,
          });
        } catch (error) {
          const isError = outputError(error as Error, path, { failOnWarning });

          if (isError) {
            errorEncountered = true;
          }
        }
      }),
    );
    if (!errorEncountered) {
      await outputResult(result, { pretty, output });
    }
  });

program.parse();
