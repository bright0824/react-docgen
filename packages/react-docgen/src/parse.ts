import Documentation from './Documentation';
import type { DocumentationObject } from './Documentation';
import postProcessDocumentation from './utils/postProcessDocumentation';
import babelParse from './babelParser';
import type { NodePath } from '@babel/traverse';
import type { Handler } from './handlers';
import type { ComponentNode } from './resolver';
import FileState from './FileState';
import type { InternalConfig } from './config';

const ERROR_MISSING_DEFINITION = 'No suitable component definition found.';

function executeHandlers(
  handlers: Handler[],
  componentDefinitions: Array<NodePath<ComponentNode>>,
): DocumentationObject[] {
  return componentDefinitions.map(
    (componentDefinition): DocumentationObject => {
      const documentation = new Documentation();

      handlers.forEach(handler => handler(documentation, componentDefinition));

      return postProcessDocumentation(documentation.toObject());
    },
  );
}

/**
 * Takes JavaScript source code and returns an object with the information
 * extract from it.
 *
 * `resolver` is a strategy to find the AST node(s) of the component
 * definition(s) inside `src`.
 * It is a function that gets passed the program AST node of
 * the source as first argument, and a reference to the parser as second argument.
 *
 * This allows you define your own strategy for finding component definitions.
 *
 * `handlers` is an array of functions which are passed a reference to the
 * component definitions (extracted by `resolver`) so that they can extract
 * information from it. They get also passed a reference to a `Documentation`
 * object to attach the information to. A reference to the parser is parsed as the
 * last argument.
 */
export default function parse(
  code: string,
  config: InternalConfig,
): DocumentationObject[] {
  const { babelOptions, handlers, importer, resolver } = config;
  const ast = babelParse(code, babelOptions);

  const fileState = new FileState(babelOptions, {
    ast,
    code,
    importer,
  });

  const componentDefinitions = resolver(fileState);

  if (componentDefinitions.length === 0) {
    throw new Error(ERROR_MISSING_DEFINITION);
  }

  return executeHandlers(handlers, componentDefinitions);
}

export { ERROR_MISSING_DEFINITION };
