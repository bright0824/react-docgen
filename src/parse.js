/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import Documentation from './Documentation';
import postProcessDocumentation from './utils/postProcessDocumentation';

import babylon from './babylon';
import recast from 'recast';

var ERROR_MISSING_DEFINITION = 'No suitable component definition found.';

function executeHandlers(handlers, componentDefinitions) {
  return componentDefinitions.map(componentDefinition => {
    var documentation = new Documentation();
    handlers.forEach(handler => handler(documentation, componentDefinition));
    return postProcessDocumentation(documentation.toObject());
  });
}

/**
 * Takes JavaScript source code and returns an object with the information
 * extract from it.
 *
 * `resolver` is a strategy to find the AST node(s) of the component
 * definition(s) inside `src`.
 * It is a function that gets passed the program AST node of
 * the source as first argument, and a reference to recast as second argument.
 *
 * This allows you define your own strategy for finding component definitions.
 *
 * `handlers` is an array of functions which are passed a reference to the
 * component definitions (extracted by `resolver`) so that they can extract
 * information from it. They get also passed a reference to a `Documentation`
 * object to attach the information to.
 *
 * If `resolver` returns an array of component definitions, `parse` will return
 * an array of documentation objects. If `resolver` returns a single node
 * instead, `parse` will return a documentation object.
 */
export default function parse(
  src: string,
  resolver: Resolver,
  handlers: Array<Handler>
): Array<Object>|Object {
  var ast = recast.parse(src, {esprima: babylon});
  var componentDefinitions = resolver(ast.program, recast);

  if (Array.isArray(componentDefinitions)) {
    if (componentDefinitions.length === 0) {
      throw new Error(ERROR_MISSING_DEFINITION);
    }
    return executeHandlers(handlers, componentDefinitions);
  } else if (componentDefinitions) {
    return executeHandlers(handlers, [(componentDefinitions: NodePath)])[0];
  }

  throw new Error(ERROR_MISSING_DEFINITION);
}

export {ERROR_MISSING_DEFINITION};
