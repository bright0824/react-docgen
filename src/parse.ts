import Documentation from './Documentation';
import type { DocumentationObject } from './Documentation';
import postProcessDocumentation from './utils/postProcessDocumentation';
import buildParser from './babelParser';
import type { Options } from './babelParser';
import type { NodePath } from '@babel/traverse';
import type { Handler } from './handlers';
import type { Importer } from './importer';
import type { Resolver } from './resolver';
import FileState from './FileState';

const ERROR_MISSING_DEFINITION = 'No suitable component definition found.';

function executeHandlers(
  handlers: Handler[],
  componentDefinitions: NodePath[],
): DocumentationObject[] {
  return componentDefinitions.map(
    (componentDefinition: NodePath): DocumentationObject => {
      const documentation = new Documentation();
      handlers.forEach(handler => handler(documentation, componentDefinition));
      return postProcessDocumentation(documentation.toObject());
    },
  );
}

/**
 * Takes JavaScript source code and returns an object with the information
 * extract from it.
 *
 * `resolver` is a strategy to find the AST node(s) of the component
 * definition(s) inside `src`.
 * It is a function that gets passed the program AST node of
 * the source as first argument, and a reference to the parser as second argument.
 *
 * This allows you define your own strategy for finding component definitions.
 *
 * `handlers` is an array of functions which are passed a reference to the
 * component definitions (extracted by `resolver`) so that they can extract
 * information from it. They get also passed a reference to a `Documentation`
 * object to attach the information to. A reference to the parser is parsed as the
 * last argument.
 */
export default function parse(
  code: string,
  resolver: Resolver,
  handlers: Handler[],
  importer: Importer,
  options: Options = {},
): DocumentationObject[] {
  const parser = buildParser(options);
  const ast = parser(code);

  const fileState = new FileState(options, {
    ast,
    code,
    importer,
    parser,
  });

  const componentDefinitions = resolver(fileState);

  if (componentDefinitions.length === 0) {
    throw new Error(ERROR_MISSING_DEFINITION);
  }
  return executeHandlers(handlers, componentDefinitions);
}

export { ERROR_MISSING_DEFINITION };
