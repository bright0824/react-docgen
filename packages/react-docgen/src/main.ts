import * as builtinHandlers from './handlers/index.js';
import parse from './parse.js';
import * as builtinResolvers from './resolver/index.js';
import {
  fsImporter,
  ignoreImporter,
  makeFsImporter,
} from './importer/index.js';
import * as utils from './utils/index.js';
import type { Documentation } from './Documentation.js';
import type DocumentationBuilder from './Documentation.js';
import type {
  Resolver,
  ResolverClass,
  ResolverFunction,
} from './resolver/index.js';
import type { Importer } from './importer/index.js';
import type { Handler } from './handlers/index.js';
import type FileState from './FileState.js';
import type { Config } from './config.js';
import { createConfig, defaultHandlers } from './config.js';
import { ERROR_CODES } from './error.js';

const builtinImporters = {
  fsImporter,
  ignoreImporter,
};

declare module '@babel/traverse' {
  export interface HubInterface {
    file: FileState;
    parse: typeof FileState.prototype.parse;
    import: typeof FileState.prototype.import;
  }

  export interface Hub {
    file: FileState;
    parse: typeof FileState.prototype.parse;
    import: typeof FileState.prototype.import;
  }
}

/**
 * Parse the *src* and scan for react components based on the config
 * that gets supplied.
 *
 * The default resolvers look for *exported* react components.
 *
 * By default all handlers are applied, so that all possible
 * different use cases are covered.
 *
 * The default importer is the fs-importer that tries to resolve
 * files based on the nodejs resolve algorithm.
 */
function defaultParse(
  src: Buffer | string,
  config: Config = {},
): Documentation[] {
  const defaultConfig = createConfig(config);

  return parse(String(src), defaultConfig);
}

export type { NodePath } from '@babel/traverse';
export type * as babelTypes from '@babel/types';

export {
  builtinHandlers,
  builtinResolvers,
  builtinImporters,
  defaultHandlers,
  makeFsImporter,
  defaultParse as parse,
  utils,
  ERROR_CODES,
};

export type {
  Importer,
  Handler,
  Resolver,
  ResolverClass,
  ResolverFunction,
  FileState,
  Config,
  Documentation,
  DocumentationBuilder,
};
