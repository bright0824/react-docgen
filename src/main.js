/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import * as handlers from './handlers';
import parse from './parse';
import * as resolver from './resolver';
import * as utils from './utils';

var defaultResolver = resolver.findExportedComponentDefinition;
var defaultHandlers = [
  handlers.propTypeHandler,
  handlers.propTypeCompositionHandler,
  handlers.propDocBlockHandler,
  handlers.flowTypeHandler,
  handlers.flowTypeDocBlockHandler,
  handlers.defaultPropsHandler,
  handlers.componentDocblockHandler,
  handlers.displayNameHandler,
  handlers.componentMethodsHandler,
  handlers.componentMethodsJsDocHandler,
];

/**
 * See `lib/parse.js` for more information about the arguments. This function
 * simply sets default values for convenience.
 *
 * The default resolver looks for *exported* `React.createClass(def)` calls
 * and expected `def` to resolve to an object expression.
 *
 * The default `handlers` look for `propTypes` and `getDefaultProps` in the
 * provided object expression, and extract prop type information, prop
 * documentation (from docblocks), default prop values and component
 * documentation (from a docblock).
 */
function defaultParse( // eslint-disable-line no-unused-vars
  src: string,
  resolver?: ?Resolver, // eslint-disable-line no-shadow
  handlers?: ?Array<Handler> // eslint-disable-line no-shadow
): Array<Object>|Object {
  if (!resolver) {
    resolver = defaultResolver;
  }
  if (!handlers) {
    handlers = defaultHandlers;
  }

  return parse(src, resolver, handlers);
}

export {
  defaultParse as parse,
  defaultHandlers,
  handlers,
  resolver,
  utils,
};
