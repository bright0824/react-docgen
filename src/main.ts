import * as allHandlers from './handlers';
import parse from './parse';
import * as allResolvers from './resolver';
import * as allImporters from './importer';
import * as utils from './utils';
import type { DocumentationObject as Documentation } from './Documentation';
import type { Resolver } from './resolver';
import type { Importer } from './importer';
import type { Handler } from './handlers';
import type FileState from './FileState';
import type { Config } from './config';
import { createConfig, defaultHandlers } from './config';

declare module '@babel/traverse' {
  export interface HubInterface {
    file: FileState;
    parse: typeof FileState.prototype.parse;
    import: typeof FileState.prototype.import;
  }

  export interface Hub {
    file: FileState;
    parse: typeof FileState.prototype.parse;
    import: typeof FileState.prototype.import;
  }
}

/**
 * Parse the *src* and scan for react components based on the config
 * that gets supplied.
 *
 * The default resolvers look for *exported* react components.
 *
 * By default all handlers are applied, so that all possible
 * different use cases are covered.
 *
 * The default importer is the fs-importer that tries to resolve
 * files based on the nodejs resolve algorithm.
 */
function defaultParse(
  src: Buffer | string,
  config: Config = {},
): Documentation[] {
  const defaultConfig = createConfig(config);

  return parse(String(src), defaultConfig);
}

export {
  defaultParse as parse,
  defaultHandlers,
  allHandlers as handlers,
  allResolvers as resolver,
  allImporters as importers,
  utils,
};

export type { Importer, Handler, Resolver, FileState, Config, Documentation };
