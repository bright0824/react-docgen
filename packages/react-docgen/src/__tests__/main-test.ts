import { handlers, parse } from '../main.js';
import { ERROR_MISSING_DEFINITION } from '../parse.js';
import { describe, expect, test } from 'vitest';

// TODO make fixtures out of them?
describe('main', () => {
  function testMain(source: string) {
    test('parses with default resolver/handlers', () => {
      const docs = parse(source);

      expect(docs).toMatchSnapshot();
    });

    test('parses with custom handlers', () => {
      const docs = parse(source, {
        handlers: [handlers.componentDocblockHandler],
      });

      expect(docs).toMatchSnapshot();
    });
  }

  describe('React.createClass', () => {
    testMain(`
      var React = require("react");
      var PropTypes = React.PropTypes;

      var defaultProps = {
        foo: true,
      };
      var propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
       * Example component description
       */
      var Component = React.createClass({
        displayName: 'ABC',
        propTypes,
        getDefaultProps: function() {
          return defaultProps;
        }
      });
      module.exports = Component
    `);
  });

  describe('Class definition', () => {
    testMain(`
      const React = require("react");
      const PropTypes = React.PropTypes;

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
       * Example component description
       */
      export default class Component extends React.Component {
        static propTypes = propTypes;
        // ...
      }
      Component.defaultProps = defaultProps;
      Component.displayName = 'ABC';
    `);
  });

  describe('Stateless Component definition: ArrowFunctionExpression', () => {
    testMain(`
      import React, {PropTypes} from "react";

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
        * Example component description
        */
      let Component = props => <div />;
      Component.displayName = 'ABC';
      Component.defaultProps = defaultProps;
      Component.propTypes = propTypes;

      export default Component;
    `);
  });

  describe('Stateless Component definition: FunctionDeclaration', () => {
    testMain(`
      import React, {PropTypes} from "react";

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
      * Example component description
      */
      function Component (props) {
        return <div />;
      }

      Component.displayName = 'ABC';
      Component.defaultProps = defaultProps;
      Component.propTypes = propTypes;

      export default Component;
    `);
  });

  describe('Stateless Component definition: FunctionExpression', () => {
    testMain(`
      import React, {PropTypes} from "react";

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
      * Example component description
      */
      let Component = function(props) {
        return React.createElement('div', null);
      }

      Component.displayName = 'ABC';
      Component.defaultProps = defaultProps;
      Component.propTypes = propTypes;

      export default Component;
    `);
  });

  describe('Stateless Component definition', () => {
    test('is not so greedy', () => {
      const source = `
        import React, {PropTypes} from "react";

        /**
        * Example component description
        */
        let NotAComponent = function(props) {
          let HiddenComponent = () => React.createElement('div', null);

          return 7;
        }

        NotAComponent.displayName = 'ABC';
        NotAComponent.defaultProps = {
            foo: true
        };

        NotAComponent.propTypes = {
          /**
          * Example prop description
          */
          foo: PropTypes.bool
        };

        export default NotAComponent;
      `;

      expect(() => parse(source)).toThrowError(ERROR_MISSING_DEFINITION);
    });
  });
});
