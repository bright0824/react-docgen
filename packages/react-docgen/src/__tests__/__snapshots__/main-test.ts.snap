// Vitest Snapshot v1

exports[`main > Class definition > parses with custom handlers 1`] = `
[
  {
    "description": "Example component description",
  },
]
`;

exports[`main > Class definition > parses with default resolver/handlers 1`] = `
[
  {
    "description": "Example component description",
    "displayName": "ABC",
    "methods": [],
    "props": {
      "foo": {
        "defaultValue": {
          "computed": false,
          "value": "true",
        },
        "description": "Example prop description",
        "required": false,
        "type": {
          "name": "bool",
        },
      },
    },
  },
]
`;

exports[`main > React.createClass > parses with custom handlers 1`] = `
[
  {
    "description": "Example component description",
  },
]
`;

exports[`main > React.createClass > parses with default resolver/handlers 1`] = `
[
  {
    "description": "Example component description",
    "displayName": "ABC",
    "methods": [],
    "props": {
      "foo": {
        "defaultValue": {
          "computed": false,
          "value": "true",
        },
        "description": "Example prop description",
        "required": false,
        "type": {
          "name": "bool",
        },
      },
    },
  },
]
`;

exports[`main > Stateless Component definition: ArrowFunctionExpression > parses with custom handlers 1`] = `
[
  {
    "description": "Example component description",
  },
]
`;

exports[`main > Stateless Component definition: ArrowFunctionExpression > parses with default resolver/handlers 1`] = `
[
  {
    "description": "Example component description",
    "displayName": "ABC",
    "methods": [],
    "props": {
      "foo": {
        "defaultValue": {
          "computed": false,
          "value": "true",
        },
        "description": "Example prop description",
        "required": false,
        "type": {
          "name": "bool",
        },
      },
    },
  },
]
`;

exports[`main > Stateless Component definition: FunctionDeclaration > parses with custom handlers 1`] = `
[
  {
    "description": "Example component description",
  },
]
`;

exports[`main > Stateless Component definition: FunctionDeclaration > parses with default resolver/handlers 1`] = `
[
  {
    "description": "Example component description",
    "displayName": "ABC",
    "methods": [],
    "props": {
      "foo": {
        "defaultValue": {
          "computed": false,
          "value": "true",
        },
        "description": "Example prop description",
        "required": false,
        "type": {
          "name": "bool",
        },
      },
    },
  },
]
`;

exports[`main > Stateless Component definition: FunctionExpression > parses with custom handlers 1`] = `
[
  {
    "description": "Example component description",
  },
]
`;

exports[`main > Stateless Component definition: FunctionExpression > parses with default resolver/handlers 1`] = `
[
  {
    "description": "Example component description",
    "displayName": "ABC",
    "methods": [],
    "props": {
      "foo": {
        "defaultValue": {
          "computed": false,
          "value": "true",
        },
        "description": "Example prop description",
        "required": false,
        "type": {
          "name": "bool",
        },
      },
    },
  },
]
`;
